### PR TITLE
Added Logging

### PR DESCRIPTION
**Pull request by georgesoteriou**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,34 +1,43 @@
...
-// setup mongoDB database
+//========================mongoDB===========================
 const db = require('./src/db.js')
+logger.info('(mongodb setup) done')
 
-//= =========================LOGIN PAGE======================
+//= =========================LOGIN PAGE=====================
 const login = require('./src/login.js')
-login.setup(app, db)
+login.setup(app, db, logger)
```
Comment (jackpordi) [inline app.js]: We shouldn't be passing around the logger object as an argument to the function.
Just do a require separately in the login.js file.

File: app.js [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
-sponsor.setup(app, db)
+sponsor.setup(app, db, logger)
+logger.info('(sponsor setup) done')
 
 //= ==========================PORTAL=========================
-
-const portal = require('./src/portal.js')
-portal.setup(app, db)
+const admin = require('./src/admin.js')
+admin.setup(app, db, logger)
```
Comment (jackpordi) [inline app.js]: Again

File: auth/auth.js [hunk 3]
```diff
@@ -2,20 +2,21 @@ const krb5 = require('node-krb5')
 const rp = require('request-promise') 
 const fs = require('fs') 
 
-const authpath = __dirname + '/auth.json' 
+const authpath = './auth.json' 
 
-exports.authSponsor = (user, pass, db, session, callback) => {
+exports.authSponsor = (user, pass, db, session,logger, callback) => {
```
Comment (jackpordi) [inline auth/auth.js]: again, no logger passing through as function arg needed

**Comment by georgesoteriou** [comment]
found a bug in admin portal

File: auth/auth.js [hunk 4]
```diff
@@ -32,21 +33,21 @@ var options = {
...
 }
 
-exports.authUser = (user, pass, session, callback) => {
+exports.authUser = (user, pass, session, logger, callback) => {
   //for debugging
-  console.log('User ' + user + ' trying to login...') 
+  logger.info('Member ' + user + ' trying to login...') 
   //KERBEROS AUTHENTICATION
-  console.log('starting Kerberos Authentication...') 
+  logger.info('starting Kerberos Authentication...')
```
Comment (jackpordi) [inline auth/auth.js]: capital S

File: auth/auth.js [hunk 5]
```diff
@@ -32,21 +33,21 @@ var options = {
...
-  console.log('User ' + user + ' trying to login...') 
+  logger.info('Member ' + user + ' trying to login...') 
   //KERBEROS AUTHENTICATION
-  console.log('starting Kerberos Authentication...') 
+  logger.info('starting Kerberos Authentication...') 
   krb5.authenticate(user + '@IC.AC.UK' , pass, (err) => {
     if(err){
       //WRONG PASSWORD/INVALID USER
-      console.log('err ' + err) 
+      logger.info('kerberos error: ' + err)
```
Comment (jackpordi) [inline auth/auth.js]: space after :

File: auth/auth.js [hunk 6]
```diff
@@ -58,44 +59,45 @@ exports.authUser = (user, pass, session, callback) => {
...
 
 //check memeber is docsoc
-checkMember = (session, user, callback) => {
+checkMember = (session, user, logger, callback) => {
   var auth = fs.readFileSync(authpath) 
   var data = JSON.parse(auth).find(el => el.Login === user) 
   if(data) {
     //VALID USER
-    console.log('User ' + user + ' successfully loged in') 
+    logger.info('User ' + user + ' successfully loged in')
```
Comment (jackpordi) [inline auth/auth.js]: 'logged'

File: auth/auth.js [hunk 7]
```diff
@@ -58,44 +59,45 @@ exports.authUser = (user, pass, session, callback) => {
...
+    session.type = 'member'
+    session.data = data
+    session.user = user
     //make folder
     callback(true)
     return
   }else{
     //NON DOCSOC USER
-    console.log('Not member of DoCSoc') 
+    logger.info('Not member of DoCSoc')
```
Comment (jackpordi) [inline auth/auth.js]: "Not a member"

File: src/db.js [hunk 8]
```diff
@@ -1,11 +1,14 @@
...
-mongoose.connect('mongodb://127.0.0.1/portal') 
+mongoose.connect('mongodb://127.0.0.1:27017/portal', {useNewUrlParser: true }) 
 var db = mongoose.connection 
-db.on('error', console.error.bind(console, 'connection error:')) 
+db.on('error', function(err) {
+  logger.error('failed to connect to mongodb: '+ err) 
+})
 db.once('open', function() {
-  console.log('connected to db') 
+  logger.info('connected to mongodb')
```
Comment (jackpordi) [inline src/db.js]: Capital C

File: src/member.js [hunk 9]
```diff
@@ -76,7 +76,10 @@ exports.setup = (app, db) => {
     }
 
     db.Sponsor.find({username: req.params.sponsor} , (err, sponsor) => {
-      if (err) return
+      if (err) {
+        logger.error(err)
```
Comment (jackpordi) [inline src/member.js]: Maybe slightly more verbose, so we have context as to where the error is occuring

File: src/member.js [hunk 10]
```diff
@@ -90,7 +93,10 @@ exports.setup = (app, db) => {
           var ext = req.files['document'+i].name.split('.').pop();
           //Save files
           req.files['document'+i].mv(path + req.body['documentname'+i] + '.' + ext, function(err) {
-            if (err) return res.status(500).send(err) 
+            if (err) {
+              logger.error(err)
```
Comment (jackpordi) [inline src/member.js]: ^

File: src/member.js [hunk 11]
```diff
@@ -102,8 +108,14 @@ exports.setup = (app, db) => {
           position.users.push(data)
         }
       });
+      
       sponsor[0].save((err, user) => {
-        if (err) return 
+        if (err) {
+          logger.error(err)
```
Comment (jackpordi) [inline src/member.js]: ^

File: src/member.js [hunk 12]
```diff
@@ -116,12 +128,12 @@ exports.setup = (app, db) => {
     //check if valid path
     var sponsorpath = './sponsors/' + req.params.sponsor + '/'
     if(!fs.existsSync(sponsorpath)){
-      console.log(req.params.sponsor+ " sponsor path magically deleted SOMETHING HAS GONE TERRIBLY WRONG or user out of sync")
+      logger.info(req.params.sponsor+ " sponsor path magically deleted SOMETHING HAS GONE TERRIBLY WRONG or user out of sync")
```
Comment (jackpordi) [inline src/member.js]: should be warning

File: src/member.js [hunk 13]
```diff
@@ -116,12 +128,12 @@ exports.setup = (app, db) => {
...
       res.redirect('/member')
     }
     var pospath = sponsorpath + req.params.posname + '/'
     if(!fs.existsSync(pospath)){
-      console.log(req.params.posname + " position, of " + req.params.sponsor +" sponsor path magically deleted SOMETHING HAS GONE TERRIBLY WRONG or user out of sync")
+      logger.info(req.params.posname + " position, of " + req.params.sponsor +" sponsor path magically deleted SOMETHING HAS GONE TERRIBLY WRONG or user out of sync")
```
Comment (jackpordi) [inline src/member.js]: ^

File: src/sponsor.js [hunk 14]
```diff
@@ -99,31 +101,35 @@ exports.setup = (app, db) => {
...
   }) 
   
   //Add new Position
   app.post('/sponsor/add-position', (req,res,next) => {
     check(req,res,next)
   },(req,res) => {
     db.Sponsor.find({username: req.session.user} , (err, sponsor) => {
-      if (err) return 
+      if (err) {         
+        logger.error('find sponsor: ' + err)
```
Comment (jackpordi) [inline src/sponsor.js]: Is this an error for failing to find a sponsor? Maybe change to "Failed to find sponsor: "

File: src/sponsor.js [hunk 15]
```diff
@@ -243,14 +282,22 @@ exports.setup = (app, db) => {
...
     db.Sponsor.find({username: req.session.user} , (err, sponsor) => {
-      if (err) return  
+      if (err) {
+        logger.error('find sponsor: ' + err)
+        return
+      }  
       if(req.body.new && sponsor[0].password === req.body.old && req.body.new === req.body.new2) {
         sponsor[0].password = req.body.new
         sponsor[0].save((err, user) => {
-          if (err) return
```
Comment (jackpordi) [inline src/sponsor.js]: we are nesting functions with an 'err' argument inside each other. Not sure about whether NodeJS will be smart enough to deal with it or not but it's still a readability nightmare and not good practice.

File: auth/auth.js (hunk 5 above)
Comment (georgesoteriou) [reply auth/auth.js]: we want that though

File: src/sponsor.js (hunk 15 above)
Comment (georgesoteriou) [reply src/sponsor.js]: shall i add different names for the errors then?

File: src/sponsor.js [hunk 16]
```diff
@@ -149,15 +150,15 @@ exports.setup = (app, db, logger) => {
...
+          if (err) {         
+            logger.error('Failed to update sponsor on adding new position: ' + err)         
+            return       
+          } 
           logger.info(req.session.user + ' succesfully added new position ' + req.body.name)
           res.redirect('/sponsor/#positions-tab-nav')
         }) 
       }else{
-        logger.info(req.session.user + ' succesfully failed to add new position ' + req.body.name)
+        logger.info(req.session.user + ' succesfully failed to add new position because name already exists' + req.body.name)
```
Comment (jackpordi) [inline src/sponsor.js]: Didn't see this the first time, but 'successfully failed'?

File: src/sponsor.js [hunk 17]
```diff
@@ -257,18 +258,18 @@ exports.setup = (app, db, logger) => {
...
-        return
-      }  
+      if (err) {         
+        logger.error('Failed to find sponsor: ' + err)         
+        return       
+      } 
       sponsor[0].info.email = req.body.email
       sponsor[0].info.description = req.body.description
       sponsor[0].info.link = req.body.link
       sponsor[0].save((err, user) => {
```
Comment (jackpordi) [inline src/sponsor.js:268]: is it sponsor[0] because mongoose returns a list of 1 back to us?

Comment (jackpordi) [reply src/sponsor.js:268]: No change required (might be worth creating a wrapper around db later down the line), just asking

Comment (georgesoteriou) [reply src/sponsor.js:268]: Yea thats exactly it

File: src/sponsor.js (hunk 16 above)
Comment (georgesoteriou) [reply src/sponsor.js]: Thats why i added the explanation of name already exists but i guess we can rephrase it

**Pull request merged**